### PR TITLE
Concept documentation for user roles, sorting and filtering

### PR DESCRIPTION
**Pull request by ikas**
Related to [GTC-733](https://gfw.atlassian.net/browse/GTC-733).

### CODE REVIEW

File: source/includes/api/_before_you_get_started.md [hunk 1]
```diff
@@ -56,6 +56,134 @@ Below you can find a list of RW API resources that use the `application` field:
...
+  "provider": "local",
+  "extraUserData": {
+    "apps": ["rw"]
+  },
+  "createdAt": 1597230329945,
+  "iat": 1597230329
+}
+```
+
+RW API users have a role associated with it, defined in the `role` field of each user. You can check your own role by consulting your user information using the `GET /users/me` endpoint or getting a JWT token and decoding its information. The `role` of the user is defined as a string, and it can take one of the following values:
```
Comment (tiagojsag) [inline source/includes/api/_before_you_get_started.md]: I'd link to the get user docs instead

File: source/includes/api/_before_you_get_started.md [hunk 2]
```diff
@@ -56,6 +56,134 @@ Below you can find a list of RW API resources that use the `application` field:
...
+
+### Role-based access control
+
+> Typical hierarchy for roles:
+
+```
+USER (least privileges) < MANAGER < ADMIN (most privileges)
+```
+
+The role field is usually used across the RW API for controlling access to API resources. A common pattern you’ll find on some services is: `USER` accounts can only create new resources, `MANAGER` accounts can create new resources, and edit or delete resources created by them, while `ADMIN` accounts can do all of the above even for resources created by other users. While not required nor enforced, typically user roles are used hierarchically, being `USER` the role with the least privileges, and `ADMIN` the one with most privileges.
```
Comment (tiagojsag) [inline source/includes/api/_before_you_get_started.md]: break this down as bulletpoints, and describe read permissions too.

File: source/includes/api/_before_you_get_started.md [hunk 3]
```diff
@@ -56,6 +56,134 @@ Below you can find a list of RW API resources that use the `application` field:
...
+
+Multiple sorting criteria can be used, separating them by commas. You can also specify the sorting order by prepending the criteria with either `-` for descending order or `+` for ascending order. By default, ascending order is assumed.
+
+Keep in mind that it’s up to each individual RW API service (dataset, widget, layer, etc) to define and implement the sorting mechanisms. Because of this, the examples above may not be true for all cases. Refer to the documentation of each resource and endpoint for more details on sorting.
+
+## Filtering
```
Comment (tiagojsag) [inline source/includes/api/_before_you_get_started.md:138]: we may want to revisit the existing sections on filtering and sorting, and DRY them by pointing here.

**Review by tiagojsag — CHANGES_REQUESTED** [review]
minor improvements needed, but otherwise LGTM

File: source/includes/api/_before_you_get_started.md (hunk 1 above)
Comment (ikas) [reply source/includes/api/_before_you_get_started.md]: Added a link to the get current user docs and removed the examples since we already have that in the get current user docs, thanks for the suggestion :) 

I did however keep the additional information about which values the `role` can have, because we don't have that info in the user docs (we don't have a model reference there) - do you think it makes sense to add a model reference to the users? If so, we can then remove this from here and just point to the model reference!

File: source/includes/api/_before_you_get_started.md (hunk 3 above)
Comment (ikas) [reply source/includes/api/_before_you_get_started.md:138]: agreed :+1:

Comment (ikas) [reply source/includes/api/_before_you_get_started.md:138]: gonna do the same for sorting

File: source/includes/api/_before_you_get_started.md (hunk 1 above)
Comment (tiagojsag) [reply source/includes/api/_before_you_get_started.md]: Having a model reference for users would make sense, but it's out of scope here, so let's not do it now.

Having the roles here: yes, I think it makes sense to have the 3 roles here, along with their description, because that will be common for all usages of roles. So, for example, we could go to dataset/widget/layer and say "there's a role param here, read the concept page for details" and the details here would hold true. Or, in other words, this section covers behavior, the endpoints cover CRUD

File: source/includes/api/_before_you_get_started.md [hunk 4]
```diff
@@ -56,6 +56,130 @@ Below you can find a list of RW API resources that use the `application` field:
...
+
+Keep in mind that it’s up to each individual RW API service (dataset, widget, layer, etc) to define how they restrict or allow actions based on these or other factors, so the examples above may not be true for all cases. Refer to the documentation of each resource and endpoint for more details on restrictions they may have regarding user accounts and their properties.
+
+### How to change the role of an user
+
+Changing role of users is restricted to `ADMIN` users, so if you intend to upgrade your user role to a `MANAGER` or `ADMIN` role, please get in touch with one of the `ADMIN` users and request the change. If you are already an `ADMIN` user and you intend to change the role of another user, you can do so using the [`PATCH /users/:id` endpoint](/index-rw.html#update-another-user-39-s-account-details).
+
+### Which services comply with these guidelines
+
+The following endpoints adhere to the Sorting conventions defined above:
```
Comment (ethanroday) [inline source/includes/api/_before_you_get_started.md]:
```suggestion
The following endpoints adhere to the user role conventions defined above:
```

File: source/includes/api/_before_you_get_started.md [hunk 5]
```diff
@@ -56,6 +56,130 @@ Below you can find a list of RW API resources that use the `application` field:
...
+curl -X GET https://api.resourcewatch.org/v1/dataset?sort=name,description
+```
+
+> Example request sorting by multiple conditions, descending and ascending:
+
+```shell
+curl -X GET https://api.resourcewatch.org/v1/dataset?sort=-name,+description
+```
+
+As a rule of thumb, you can sort RW API resources using the `sort` query parameter. Usually, sorting can be performed using any field from the resource schema, so be sure to check each resource's model reference to find which fields can be used for sorting. In some cases, you also have the possibility of sorting by fields that are not present in the resource model (e.g., when fetching datasets, you can sort by `user.name` and `user.role` to sort datasets by the name or role of the owner of the dataset) - check each resource's documentation to find out which additional sorting criteria you have available. Sorting by nested fields is not supported at the moment.
```
Comment (ethanroday) [inline source/includes/api/_before_you_get_started.md]: That nested fields note is confusing right after talking about `user.name`. Maybe just say: "Sorting by nested fields is not generally supported, but may be implemented for particular resources."

File: source/includes/api/_before_you_get_started.md [hunk 6]
```diff
@@ -56,6 +56,130 @@ Below you can find a list of RW API resources that use the `application` field:
...
+
+> Example request filtering by an array field using the `@` AND multi-value separator:
+
+```shell
+curl -X GET https://api.resourcewatch.org/v1/dataset?application=rw@gfw
+```
+
+Like in the case of sorting, most RW API resources allow filtering the returned results of list endpoints using query parameters. As a rule of thumb, you can use the API resource's fields as query parameter filters, as shown in the examples on the side. You can also combine different query parameters into a complex `and` logic filter.
+
+String type fields support and expect a regex value, unless detailed otherwise on documentation for each API resource. Although typically they will match exact strings, you may have to escape certain characters (PCRE v8.42 spec).
```
Comment (ethanroday) [inline source/includes/api/_before_you_get_started.md]: This is a little confusing. So is it a regex value or an exact string?

File: source/includes/api/_before_you_get_started.md [hunk 7]
```diff
@@ -56,6 +56,130 @@ Below you can find a list of RW API resources that use the `application` field:
...
+curl -X GET https://api.resourcewatch.org/v1/dataset?application=rw,gfw
+```
+
+> Example request filtering by an array field using the `@` AND multi-value separator:
+
+```shell
+curl -X GET https://api.resourcewatch.org/v1/dataset?application=rw@gfw
+```
+
+Like in the case of sorting, most RW API resources allow filtering the returned results of list endpoints using query parameters. As a rule of thumb, you can use the API resource's fields as query parameter filters, as shown in the examples on the side. You can also combine different query parameters into a complex `and` logic filter.
```
Comment (ethanroday) [inline source/includes/api/_before_you_get_started.md]: What about OR with comma for single-valued fields? Like https://api.resourcewatch.org/v1/dataset?name=viirs,modis to get datasets matching viirs or modis. Should explicitly call out support or lack thereof.

**Review by ethanroday — CHANGES_REQUESTED** [review]
Some minor changes, but otherwise looks good.

File: source/includes/api/_before_you_get_started.md (hunk 6 above)
Comment (tiagojsag) [reply source/includes/api/_before_you_get_started.md]: It's complicated. All string filters are passed to MongoDB as case-insensitive regexes: https://github.com/resource-watch/dataset/blob/a0ec9cd728df600f1209f7fb0eee90b65cfb35d3/src/services/dataset.service.js#L128

However, from my experiments, it seems that MongoDB is "smart" enough to produce the results you would expect if you were to use this filter as a string, rather than a regex. That, of course, will come with its flaws, so we can't really promise it will work for all strings as-is - hence the "you may need to escape stuff". 

So, long story short, I see two paths:
- we tell users that they may be able to get away with simple strings in most cases, but that they may have to deal/may be able to use regexes in some cases. This is the current text
- we tell users "always use regexes", which is truer to spec, and arguably the safer approach for us, but also is technically a more complex solution (I don't know about you, but the word "regexes" induces some anxiety in me).

File: source/includes/api/_before_you_get_started.md (hunk 5 above)
Comment (ikas) [reply source/includes/api/_before_you_get_started.md]: I've accepted your suggestion, and I've reworded the paragraph a bit to make it less confusing :)

File: source/includes/api/_before_you_get_started.md (hunk 7 above)
Comment (ikas) [reply source/includes/api/_before_you_get_started.md]: That's implicitly supported, as we were discussing in the comment above, by using a regex like `?name=viirs|modis` - do you think we need to explicitly highlight this as allowed?

File: source/includes/api/_before_you_get_started.md (hunk 6 above)
Comment (ethanroday) [reply source/includes/api/_before_you_get_started.md]: How about this?

```suggestion
For string type fields, the filter you pass will be interpreted as a regular expression, _not_ as a simple substring filter. This gives you greater flexibility in your search capabilities. However, it means that, if you intend to search by substring, you must escape any regex special characters in the string.
```

File: source/includes/api/_before_you_get_started.md (hunk 7 above)
Comment (ethanroday) [reply source/includes/api/_before_you_get_started.md]: I would just briefly call it out here. Something like: "Note that you can achieve a logical `or` by passing a regular expression with two disjoint options, like this: `?name=<substr_a>|<substr_b>`."

**Review by ethanroday — APPROVED** [review]
@ikas see a couple of suggestions above. Otherwise this is good to go. No need to kick it back to me for review again.
